src/render/tutorials/api-gateway-security/2_0.html.md

```diff
@@ -0,0 +1,913 @@
+    }
+
+    return webClient.postAbs(TOKEN_URL)
+        .setQueryParam("code", code)
+        .setQueryParam("client_id", clientId)
+        .setQueryParam("client_secret", clientSecret)
+        .setQueryParam("redirect_uri", redirectUri)
+        .setQueryParam("grant_type", "authorization_code")
+        .putHeader("Content-Length", "0");
+  }
+
+  private Single<HttpResponse<Buffer>> fetchUserInfo(String accessToken) {
+    return webClient.getAbs(USERINFO_URL)
+        .bearerTokenAuthentication(accessToken)
+        .rxSend();
+  }
+
+}
+
+```
+
+Let's take a look at this file step by step:
+
+1. The constructor is quite straightforward, we create a new [`webClient`](https://vertx.io/docs/vertx-web-client/java/) and save the config for later use.
+2. When a request hits our handler, the `handle` method is invoked. We try to obtain the authorization code from the request.
+3. Then we try to `exchangeCodeForToken`. The first step is to `prepareTokenRequest`. This time we have to provide the `code`, `client_id`, `client_secret`, `redirect_uri`, `grant_type` parameters and `Content-Length` header (this will be a `POST` request). For more details on those parameters see [this documentation](https://developers.google.com/identity/protocols/OAuth2WebServer#exchange-authorization-code). We send the request using `rxSend` method (for more info about Vert.x's RxJava 2 API check [this documentation](https://vertx.io/docs/vertx-web-client/java/#_rxjava_2_api)) and map the response to the acquired access token.
+4. We map the resulting observable with the token to `fetchUserInfo` which creates a `GET` request to Google API with the bearer token authentication and sends it.
+5. Finally we subscribe for the whole processing chain and end our response with a personalized *hello* message!
+
+As always, let's not forget to register our factories in the `resources/META_INF/services` directory!
+
+`io.knotx.server.api.handler.RoutingHandlerFactory`
+```
+io.knotx.examples.security.routing.GoogleLoginRedirectHandlerFactory
+io.knotx.examples.security.routing.OAuth2HelloHandlerFactory
+```
+
+### Build & Run
+
+Now let's see how it works.
+
+First, let's build the project (as mentioned before, remember to stop all the running containers!):
+
+```
+$ gradlew clean build-docker
+```
+
+and run it:
+
+```
+$ docker run -p8092:8092 knotx-example/secure-api-gateway
+```
+
+Now let's type in `http://localhost:8092/api/secure/login` in your browser.
+
+![Select Google Account](/img/blog/api-gateway-security/security-google-select-account.png)
+
+Select your account.
+
+![Select Google Account](/img/blog/api-gateway-security/security-google-hello-message.png)
+
+And there is our personalized hello message!
+
+## Summary
+
+In this tutorial we have successfully secured our API Gateway with various authentication methods. We even managed to access secured data from external API using OAuth2 flow.
```

Comment:
I think that summary deserves few more sentences after so many happened above.
Maybe write about:
- Security Handlers as way of enriching API with security in Knot.x (using SPI)
- Connecting external security provider with API and Server routing.
- Mention about importance of tests.

Reply:
Splitted into separate tutorials and added adequate summary there